morango/sync/syncsession.py

```diff
@@ -127,6 +129,14 @@ def urlresolve(self, endpoint, lookup=None):
         return url
 
     def create_sync_session(self, client_cert, server_cert, chunk_size=500):
+        # since a sync session can be used for pulling and pushing, we'll cache session objects
+        # we create so a push and pull to the same server get the same session
+        cache_key = "session_{}_{}".format(client_cert.id, server_cert.id)
+        existing_session = self.sync_sessions.get(cache_key, None)
+        if existing_session is not None:
```

Comment:
Should we be closing the sync session when it fails?

Reply:
Should we always mark the sync session closed if it fails?

Reply:
Depends on the type of failure. I believe the intended behavior is that for network failures, it wouldn't, so it could be resumed later, but for 4xx or 5xx type errors it would close it out.

But that's separate from what I was saying here: My worry was that we cache a session, and then it gets closed, but we didn't clear it from this cache, and all we're checking for here is that it isn't `None`, not that it still works. And if we then try to use it and it fails, we may _still_ not clear it from the cache, so it would then be stuck. Or I may have missed where that would be handled.